WordPress/gutenberg
# Blocks: Render columns as nested column (wrapper)

Closes #5351

This pull request seeks to address an issue where the current Columns block implementation using grid is insufficient for supporting columns of varying height where it is expected for blocks within a column to stack atop each other, thus resulting in undesirable empty space.

Before|After
---|---
![Before](https://user-images.githubusercontent.com/1779930/41172752-d45101b6-6b22-11e8-9619-4ade4c3d4315.png)|![After](https://user-images.githubusercontent.com/1779930/41172652-83195ff0-6b22-11e8-98eb-c417771c4c14.png)

**Implementation notes:**

This explores one option of addressing #5351 which was considered at https://github.com/WordPress/gutenberg/issues/5351#issuecomment-375795651, where in order to introduce a wrapper element for each column, the columns themselves are an intermediary block. This is possible through new APIs for `supportedBlocks` of a block type (#6753) and `allowedBlocks` and `template` of an `InnerBlocks` (#6546). Thus, a "Column" block is only valid within a Columns wrapper, and conversely a Column block is the only valid child of a Columns block. This will require that the Columns block fully manages its own template, which is only partially implemented thus far.

This may allow for the eventual removal of `layouts` concept altogether.

The primary alternative here is to keep the distinction of "grouped" and "ungrouped" layouts ([documentation](https://github.com/WordPress/gutenberg/tree/master/editor/components/inner-blocks)), and automatically insert a wrapper element for grouped layouts. This is challenging both in the further inconsistency between the two types of layouts, and in the current unavailability of `layouts` anywhere outside of the `edit` implementation where it is passed (it would likely need to become a top-level property of the block).

**Testing instructions:**

Verify that you can create columns, insert blocks within (via slash insertion), and that within columns, the individual blocks stack neatly atop each other when previewed on the front of your site.
